sdk/core/azure-core-amqp/src/models/amqp_properties.cpp

```diff
@@ -0,0 +1,442 @@
+    return Azure::Core::Amqp::Models::AmqpValue::Serialize(propertiesAsValue);
+  }
+
+  MessageProperties MessageProperties::Deserialize(uint8_t const* data, size_t size)
+  {
+    AmqpValue value{AmqpValue::Deserialize(data, size)};
+    PROPERTIES_HANDLE handle;
+    if (amqpvalue_get_properties(value, &handle))
+    {
+      throw std::runtime_error("Could not convert value to AMQP Properties.");
+    }
+    UniquePropertiesHandle uniqueHandle{handle};
+    handle = nullptr;
+    return _internal::MessagePropertiesFactory::FromUamqp(uniqueHandle);
+  }
+
+  namespace {
+    std::string timeToString(std::chrono::system_clock::time_point t)
+    {
+      std::time_t time = std::chrono::system_clock::to_time_t(t);
+      std::string time_str = std::ctime(&time);
+      time_str.resize(time_str.size() - 1);
+      return time_str;
+    }
+
+    size_t LogRawData(std::ostream& os, size_t startOffset, const uint8_t* const pb, size_t cb)
+    {
+      // scratch buffer which will hold the data being logged.
+      std::stringstream ss;
+
+      size_t bytesToWrite = (cb < 0x10 ? cb : 0x10);
+
+      ss << std::hex << std::right << std::setw(8) << std::setfill('0') << startOffset << ": ";
+
+      // Write the buffer data out.
+      for (size_t i = 0; i < bytesToWrite; i += 1)
+      {
+        ss << std::hex << std::right << std::setw(2) << std::setfill('0') << static_cast<int>(pb[i])
+           << " ";
+      }
+
+      // Now write the data in string format (similar to what the debugger does).
+      // Start by padding partial lines to a fixed end.
+      for (size_t i = bytesToWrite; i < 0x10; i += 1)
+      {
+        ss << "   ";
+      }
+      ss << "  * ";
+      for (size_t i = 0; i < bytesToWrite; i += 1)
+      {
+        if (isprint(pb[i]))
+        {
+          ss << pb[i];
+        }
+        else
+        {
+          ss << ".";
+        }
+      }
+      for (size_t i = bytesToWrite; i < 0x10; i += 1)
+      {
+        ss << " ";
+      }
+
+      ss << " *";
+
+      os << ss.str();
+
+      return bytesToWrite;
+    }
+  } // namespace
+
+  std::ostream& operator<<(std::ostream& os, MessageProperties const& properties)
+  {
+    os << "MessageProperties {";
+    {
+      if (properties.MessageId.HasValue())
+      {
+        os << "MessageId: " << properties.MessageId.Value();
+      }
+      else
+      {
+        os << "MessageId: <null>";
+      }
+    }
+    {
+      if (properties.UserId.HasValue())
+      {
+        os << ", UserId: ";
+        const uint8_t* pb = properties.UserId.Value().data();
+        size_t cb = properties.UserId.Value().size();
+        size_t currentOffset = 0;
+        do
+        {
+          auto cbLogged = LogRawData(os, currentOffset, pb, cb);
+          pb += cbLogged;
+          cb -= cbLogged;
+          currentOffset += cbLogged;
+          if (cb)
+          {
+            os << std::endl;
+          }
+        } while (cb);
+      }
+    }
+    if (properties.To.HasValue())
```

Comment:
Ideally, this would be a `foreach`, but it is not as easy as with the one that I proposed for `operator ==` and `ShouldSerialize`, but maybe something can be done about it as well? The goal is - you update one place, and everything starts working.

Reply:
I'm not sure I understand the construct you're proposing here. I'm not aware of any C++ constructs that allow you to iterate over the fields of a structure without first understanding the layout of the structure.